### PR TITLE
Add typings

### PR DESCRIPTION
**Pull request by resynth1943**
I added some TypeScript type definitions so I can actually use this module :P

### CODE REVIEW

**Comment by resynth1943** [comment]
That failing check has nothing to do with my PR, by the way

**Comment by koddsson** [comment]
Hey @resynth1943 and thanks for the PR.

I believe that typescript includes fetch typings out of the box regardless if you are using the polyfill or not. Adding `'dom'` to the `'lib'` key in your `tsconfig.json` should solve this for you without us needing to define types in the polyfill codebase.

Reference: https://www.typescriptlang.org/docs/handbook/compiler-options.html

**Comment by resynth1943** [comment]
Yeah, but you have to declare the module as exporting those symbols. I'll edit this PR to use the "fetch" from the TypeScript libraries, but you'll still need an export. Give me a few hours, and I'll take a look.

Regards,
Will

**Comment by resynth1943** [comment]
How do you like *this*? I see no errors from Vim lol, so I guess that means everything's OK now?

**Comment by MicahZoltu** [comment]
@koddsson You should only include `dom` `lib` if you actually are in an environment that has the full `dom` API.  If you are in NodeJS and you bring in this library, you do _not_ have the full `dom` API, so you should not add `dom` to your `lib` in `tsconfig.json`.

File: index.d.ts [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+    method?: string
+    redirect?: RequestRedirect
+    signal?: AbortSignal | null
+
+    // node-fetch extensions
+    agent?: Agent | ((parsedUrl: URL) => Agent) // =null http.Agent instance, allows custom proxy, certificate etc.
+    compress?: boolean // =true support gzip/deflate content encoding. false to disable
+    follow?: number // =20 maximum redirect count. 0 to not follow redirect
+    size?: number // =0 maximum response body size in bytes. 0 to disable
+    timeout?: number // =0 req/res timeout in ms, it resets on redirect. 0 to disable (OS limit applies)
```
Comment (MicahZoltu) [inline index.d.ts]: It looks like this was copied from `node-fetch` definitions.  I don't believe this library has any of the `node-fetch` extensions so they should not be included.

Comment (MicahZoltu) [reply index.d.ts]: Same for all of the node-fetch extensions throughout.

Comment (resynth1943) [reply index.d.ts]: Changing it now :D

File: index.d.ts [hunk 2]
```diff
@@ -0,0 +1,209 @@
+declare module 'whatwg-fetch' {
+  import {Agent} from 'http'
+  import {URLSearchParams, URL} from 'url'
```
Comment (MicahZoltu) [inline index.d.ts]: Once the node-fetch extensions get dropped, I bet you can delete these imports.  You should only need native JS types to create this declaration file.

File: index.d.ts [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+    slice(start?: number, end?: number): Blob
+  }
+
+  export class Body {
+    constructor(body?: any, opts?: {size?: number; timeout?: number})
+    arrayBuffer(): Promise<ArrayBuffer>
+    blob(): Promise<Buffer>
+    body: NodeJS.ReadableStream
+    bodyUsed: boolean
+    buffer(): Promise<Buffer>
```
Comment (MicahZoltu) [inline index.d.ts]: You should verify, but I don't think this library uses `Buffer` or `NodeJS.ReadableStream`.  I suspect this is an artifact of copying from `node-fetch` definition.

File: index.d.ts (hunk 2 above)
Comment (resynth1943) [reply index.d.ts]: Done.

**Comment by JakeChampion** [comment]
> @koddsson You should only include `dom` `lib` if you actually are in an environment that has the full `dom` API. If you are in NodeJS and you bring in this library, you do _not_ have the full `dom` API, so you should not add `dom` to your `lib` in `tsconfig.json`.

The [third bullet-point in the README.md](https://github.com/github/fetch#read-this-first) says: 
This project doesn't work under Node.js environments. It's meant for web browsers only.

This project is for browser environments, which means you would need to configure TypeScript to bring in the DOM API, which includes the `fetch` typings.

**Comment by koddsson** [comment]
> How do you like _this_?

Like I said I believe that TypeScript supports fetch typings natively so there should be any need for typings so the issue isn't the types at all.

With that being said, we don't plan on adding any type support to the fetch polyfill. They should exists in the type system already or as a 3rd party type module.

**Comment by resynth1943** [comment]
@koddsson 
```
Cannot export 'fetch'. Only local declarations can be exported from a module.ts(2661)
```

Apparently you're *wrong*.

Since you're refusing to merge this, I'll try and get them merged into DefinitelyTyped, just so this wasn't a complete waste of time. 

These typings are a necessity for TypeScript users, and I'm more than happy to spend time with this.

Many thanks,
resynth1943

**Pull request closed without merging**